File: codescmp.pas
{$I STDINC.PAS}
{$IFDEF SEGATTRS}
{$C Moveable DemandLoad Discardable }
{$ENDIF}

        UNIT CodeSCMP;

{ AS Codegeneratormodul SC/MP }

INTERFACE
        Uses NLS,
             Chunks,AsmDef,AsmSub,AsmPars,CodePseu;


IMPLEMENTATION

CONST
   FixedOrderCnt=21;
   ImmOrderCnt=8;
   RegOrderCnt=3;
   MemOrderCnt=8;
   JmpOrderCnt=4;

TYPE
   FixedOrder=RECORD
               Name:String[5];
               Code:Byte;
              END;

   FixedOrderArray=ARRAY[0..FixedOrderCnt-1] OF FixedOrder;
   ImmOrderArray=ARRAY[0..ImmOrderCnt-1] OF FixedOrder;
   RegOrderArray=ARRAY[0..RegOrderCnt-1] OF FixedOrder;
   MemOrderArray=ARRAY[0..MemOrderCnt-1] OF FixedOrder;
   JmpOrderArray=ARRAY[0..JmpOrderCnt-1] OF FixedOrder;

VAR
   CPUSCMP:CPUVar;

   FixedOrders:^FixedOrderArray;
   ImmOrders:^ImmOrderArray;
   RegOrders:^RegOrderArray;
   MemOrders:^MemOrderArray;
   JmpOrders:^JmpOrderArray;

{-----------------------------------------------------------------------------}

        PROCEDURE InitFields;
VAR
   z:Integer;

        PROCEDURE AddFixed(NName:String; NCode:Byte);
BEGIN
   IF z>=FixedOrderCnt THEN Halt(255);
   WITH FixedOrders^[z] DO
    BEGIN
     Name:=NName;
     Code:=NCode;
    END;
   Inc(z);
END;

        PROCEDURE AddImm(NName:String; NCode:Byte);
BEGIN
   IF z>=ImmOrderCnt THEN Halt(255);
   WITH ImmOrders^[z] DO
    BEGIN
     Name:=NName;
     Code:=NCode;
    END;
   Inc(z);
END;

        PROCEDURE AddReg(NName:String; NCode:Byte);
BEGIN
   IF z>=RegOrderCnt THEN Halt(255);
   WITH RegOrders^[z] DO
    BEGIN
     Name:=NName;
     Code:=NCode;
    END;
   Inc(z);
END;

        PROCEDURE AddMem(NName:String; NCode:Byte);
BEGIN
   IF z>=MemOrderCnt THEN Halt(255);
   WITH MemOrders^[z] DO
    BEGIN
     Name:=NName;
     Code:=NCode;
    END;
   Inc(z);
END;

        PROCEDURE AddJmp(NName:String; NCode:Byte);
BEGIN
   IF z>=JmpOrderCnt THEN Halt(255);
   WITH JmpOrders^[z] DO
    BEGIN
     Name:=NName;
     Code:=NCode;
    END;
   Inc(z);
END;

BEGIN
   New(FixedOrders); z:=0;
   AddFixed('LDE' ,$40); AddFixed('XAE' ,$01); AddFixed('ANE' ,$50);
   AddFixed('ORE' ,$58); AddFixed('XRE' ,$60); AddFixed('DAE' ,$68);
   AddFixed('ADE' ,$70); AddFixed('CAE' ,$78); AddFixed('SIO' ,$19);
   AddFixed('SR'  ,$1c); AddFixed('SRL' ,$1d); AddFixed('RR'  ,$1e);
   AddFixed('RRL' ,$1f); AddFixed('HALT',$00); AddFixed('CCL' ,$02);
   AddFixed('SCL' ,$03); AddFixed('DINT',$04); AddFixed('IEN' ,$05);
   AddFixed('CSA' ,$06); AddFixed('CAS' ,$07); AddFixed('NOP' ,$08);

   New(ImmOrders); z:=0;
   AddImm('LDI' ,$c4); AddImm('ANI' ,$d4); AddImm('ORI' ,$dc);
   AddImm('XRI' ,$e4); AddImm('DAI' ,$ec); AddImm('ADI' ,$f4);
   AddImm('CAI' ,$fc); AddImm('DLY' ,$8f);

   New(RegOrders); z:=0;
   AddReg('XPAL',$30); AddReg('XPAH',$34); AddReg('XPPC',$3c);

   New(MemOrders); z:=0;
   AddMem('LD'  ,$c0); AddMem('ST'  ,$c8); AddMem('AND' ,$d0);
   AddMem('OR'  ,$d8); AddMem('XOR' ,$e0); AddMem('DAD' ,$e8);
   AddMem('ADD' ,$f0); AddMem('CAD' ,$f8);

   New(JmpOrders); z:=0;
   AddJmp('JMP' ,$90); AddJmp('JP'  ,$94); AddJmp('JZ'  ,$98);
   AddJmp('JNZ' ,$9c);
END;

        PROCEDURE DeinitFields;
BEGIN
   Dispose(FixedOrders);
   Dispose(ImmOrders);
   Dispose(RegOrders);
   Dispose(MemOrders);
   Dispose(JmpOrders);
END;

{-----------------------------------------------------------------------------}

	FUNCTION DecodeReg(Asc:String; VAR Erg:Byte):Boolean;
BEGIN
   DecodeReg:=False;

   IF (Length(Asc)<>2) OR (UpCase(Asc[1])<>'P') THEN Exit;

   CASE UpCase(Asc[2]) OF
   'C':Erg:=0;
   '0'..'3':Erg:=Ord(Asc[2])-AscOfs;
   ELSE Exit;
   END;

   DecodeReg:=True;
END;

	FUNCTION DecodeAdr(Asc:String; MayInc:Boolean; PCDisp:Byte; VAR Arg:Byte):Boolean;
VAR
   Disp:Integer;
   PCVal:Word;
   Reg:Byte;
   OK:Boolean;
BEGIN
   DecodeAdr:=False;

   IF (Length(Asc)>=4) AND (Asc[Length(Asc)]=')') AND (Asc[Length(Asc)-3]='(')
      AND (DecodeReg(Copy(Asc,Length(Asc)-2,2),Arg)) THEN
    BEGIN
     Delete(Asc,Length(Asc)-3,4);
     IF Asc[1]='@' THEN
      BEGIN
       IF NOT MayInc THEN
        BEGIN
         WrError(1350); Exit;
        END;
       Delete(Asc,1,1); Inc(Arg,4);
      END;
     IF NLS_StrCaseCmp(Asc,'E')=0 THEN BAsmCode[1]:=$80
     ELSE IF Arg=0 THEN
      BEGIN
       WrXError(1445,Copy(Asc,Length(Asc)-2,2)); Exit;
      END
     ELSE
      BEGIN
       BAsmCode[1]:=EvalIntExpression(Asc,SInt8,OK);
       IF NOT OK THEN Exit;
      END;
     DecodeAdr:=True; Exit;
    END;

   PCVal:=(EProgCounter AND $f000)+((EProgCounter+1) AND $fff);
   Disp:=EvalIntExpression(Asc,UInt16,OK)-PCDisp-PCVal;
   IF OK THEN
    IF (NOT SymbolQuestionable) AND ((Disp<-128) OR (Disp>127)) THEN WrError(1370)
    ELSE
     BEGIN
      BAsmCode[1]:=Disp AND $ff; Arg:=0; DecodeAdr:=True;
     END;
END;

{-----------------------------------------------------------------------------}

	FUNCTION DecodePseudo:Boolean;
BEGIN
   DecodePseudo:=True;

   DecodePseudo:=False;
END;

        PROCEDURE ChkPage;
BEGIN
   IF ((EProgCounter) AND $f000)<>((EProgCounter+CodeLen) AND $f000) THEN WrError(250);
END;

        PROCEDURE MakeCode_SCMP;
	Far;
VAR
   z:Integer;
   OK:Boolean;
BEGIN
   CodeLen:=0; DontPrint:=False;

   { zu ignorierendes }

   if Memo('') THEN Exit;

   { Pseudoanweisungen }

   IF DecodePseudo THEN Exit;

   IF DecodeIntelPseudo(False) THEN Exit;

   { ohne Argument }

   FOR z:=0 TO FixedOrderCnt-1 DO
    WITH FixedOrders^[z] DO
     IF Memo(Name) THEN
      BEGIN
       IF ArgCnt<>0 THEN WrError(1110)
       ELSE
        BEGIN
         BAsmCode[0]:=Code; CodeLen:=1;
        END;
       Exit;
      END;

   { immediate }

   FOR z:=0 TO ImmOrderCnt-1 DO
    WITH ImmOrders^[z] DO
     IF Memo(Name) THEN
      BEGIN
       IF ArgCnt<>1 THEN WrError(1110)
       ELSE
        BEGIN
         BAsmCode[1]:=EvalIntExpression(ArgStr[1],Int8,OK);
         IF OK THEN
          BEGIN
           BAsmCode[0]:=Code; CodeLen:=2; ChkPage;
          END;
        END;
       Exit;
      END;

   { ein Register }

   FOR z:=0 TO RegOrderCnt-1 DO
    WITH RegOrders^[z] DO
     IF Memo(Name) THEN
      BEGIN
       IF ArgCnt<>1 THEN WrError(1110)
       ELSE IF NOT DecodeReg(ArgStr[1],BAsmCode[0]) THEN WrXError(1445,ArgStr[1])
       ELSE
        BEGIN
         Inc(BAsmCode[0],Code); CodeLen:=1;
        END;
       Exit;
      END;


   { ein Speicheroperand }

   FOR z:=0 TO MemOrderCnt-1 DO
    WITH MemOrders^[z] DO
     IF Memo(Name) THEN
      BEGIN
       IF ArgCnt<>1 THEN WrError(1110)
       ELSE IF DecodeAdr(ArgStr[1],True,0,BAsmCode[0]) THEN
        BEGIN
         Inc(BAsmCode[0],Code); CodeLen:=2; ChkPage;
        END;
       Exit;
      END;

   FOR z:=0 TO JmpOrderCnt-1 DO
    WITH JmpOrders^[z] DO
     IF Memo(Name) THEN
      BEGIN
       IF ArgCnt<>1 THEN WrError(1110)
       ELSE IF DecodeAdr(ArgStr[1],False,1,BAsmCode[0]) THEN
        BEGIN
         Inc(BAsmCode[0],Code); CodeLen:=2; ChkPage;
        END;
       Exit;
      END;

   IF (Memo('ILD')) OR (Memo('DLD')) THEN
    BEGIN
     IF ArgCnt<>1 THEN WrError(1110)
     ELSE IF DecodeAdr(ArgStr[1],False,0,BAsmCode[0]) THEN
      BEGIN
       Inc(BAsmCode[0],$a8+(Ord(Memo('DLD')) SHL 4)); CodeLen:=2; ChkPage;
      END;
     Exit;
    END;

   WrXError(1200,OpPart);
END;

        FUNCTION ChkPC_SCMP:Boolean;
	Far;
VAR
   ok:Boolean;
BEGIN
   CASE ActPC OF
   SegCode  : ok:=ProgCounter<$10000;
   ELSE ok:=False;
   END;
   ChkPC_SCMP:=(ok) AND (ProgCounter>=0);
END;


        FUNCTION IsDef_SCMP:Boolean;
	Far;
BEGIN
   IsDef_SCMP:=False;
END;

        PROCEDURE SwitchFrom_SCMP;
        Far;
BEGIN
   DeinitFields;
END;

        PROCEDURE SwitchTo_SCMP;
	Far;
BEGIN
   TurnWords:=False; ConstMode:=ConstModeC; SetIsOccupied:=False;

   PCSymbol:='$'; HeaderID:=$6e; NOPCode:=$08;
   DivideChars:=','; HasAttrs:=False;

   ValidSegs:=[SegCode,SegData];
   Grans[SegCode]:=1; ListGrans[SegCode]:=1; SegInits[SegCode]:=0;

   MakeCode:=MakeCode_SCMP; ChkPC:=ChkPC_SCMP; IsDef:=IsDef_SCMP;
   SwitchFrom:=SwitchFrom_SCMP; InitFields;
END;

BEGIN
   CPUSCMP:=AddCPU('SC/MP',SwitchTo_SCMP);
END.

File: codescmp.c
/* codescmp.c */
/*****************************************************************************/
/* SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only                     */
/*                                                                           */
/* AS-Portierung                                                             */
/*                                                                           */
/* Codegenerator National SC/MP                                              */
/*                                                                           */
/*****************************************************************************/

#include "stdinc.h"
#include <ctype.h>
#include <string.h>

#include "nls.h"
#include "strutil.h"
#include "chunks.h"
#include "asmdef.h"
#include "asmsub.h"
#include "asmpars.h"
#include "asmitree.h"
#include "asmallg.h"
#include "intpseudo.h"
#include "codevars.h"
#include "codepseudo.h"
#include "errmsg.h"

#include "codescmp.h"

/*---------------------------------------------------------------------------*/

static CPUVar CPUSCMP;

/*---------------------------------------------------------------------------*/

/*!------------------------------------------------------------------------
 * \fn     DecodeReg(const char *pArg, Byte *pResult)
 * \brief  check whether argument is a CPU register
 * \param  pArg source argument
 * \param  pResult resulting register # if yes
 * \return True if argument is a register
 * ------------------------------------------------------------------------ */

static Boolean DecodeReg(const char *pArg, Byte *pResult)
{
  int l = strlen(pArg);

  switch (l)
  {
    case 2:
      if (as_toupper(*pArg) != 'P')
        return False;
      pArg++;
      /* fall-through */
    case 1:
      switch (as_toupper(*pArg))
      {
        case '0':
        case '1':
        case '2':
        case '3':
          *pResult = *pArg - '0';
          return True;
        case 'C':
          *pResult = 0;
          return (l == 2);
      }
  }
  return False;
}

static Boolean DecodeAdr(const tStrComp *pArg, Boolean MayInc, Byte PCDisp, Byte *Arg)
{
  Word Target;
  Boolean OK;
  int l, SplitPos;
  tSymbolFlags Flags;

  if (((SplitPos = FindDispBaseSplit(pArg->Str, &l)) >= 0) && (l >= 4))
  {
    tStrComp Left, Right;

    StrCompSplitRef(&Left, &Right, pArg, pArg->Str + SplitPos);
    StrCompShorten(&Right, 1);
    if (DecodeReg(Right.Str, Arg))
    {
      if (*Left.Str == '@')
      {
        if (!MayInc)
        {
          WrError(ErrNum_InvAddrMode);
          return False;
        }
        StrCompIncRefLeft(&Left, 1);
        *Arg += 4;
      }
      if (!as_strcasecmp(Left.Str, "E"))
        BAsmCode[1] = 0x80;
      else if (*Arg == 0)
      {
        WrStrErrorPos(ErrNum_InvReg, &Right);
        return False;
      }
      else
      {
        BAsmCode[1] = EvalStrIntExpression(&Left, SInt8, &OK);
        if (!OK)
          return False;
      }
      return True;
    }
    else pArg->Str[l - 1] = ')';
  }

  /* no carry in PC from bit 11 to 12; additionally handle preincrement */

  Target = EvalStrIntExpressionWithFlags(pArg, UInt16, &OK, &Flags);
  if (OK)
  {
    Word PCVal = (EProgCounter() & 0xf000) + ((EProgCounter() + 1 + PCDisp) & 0xfff);
    Word Disp = (Target - PCVal) & 0xfff;

    if (mSymbolQuestionable(Flags))
      Target = PCVal;

    if (!ChkSamePage(Target, PCVal, 12, Flags));
    else if ((Disp > 0x7f) && (Disp < 0xf80)) WrError(ErrNum_JmpDistTooBig);
    else
    {
      BAsmCode[1] = Disp & 0xff;
      *Arg = 0;
      return True;
    }
  }
  return False;
}

static void ChkPage(void)
{
  if (((EProgCounter()) & 0xf000) != ((EProgCounter() + CodeLen) & 0xf000))
    WrError(ErrNum_PageCrossing);
}

/*---------------------------------------------------------------------------*/

static void DecodeFixed(Word Index)
{
  if (ChkArgCnt(0, 0))
  {
    BAsmCode[0] = Index; CodeLen = 1;
  }
}

static void DecodeImm(Word Index)
{
  if (ChkArgCnt(1, 1))
  {
    Boolean OK;

    BAsmCode[1] = EvalStrIntExpression(&ArgStr[1], Int8, &OK);
    if (OK)
    {
      BAsmCode[0] = Index; CodeLen = 2; ChkPage();
    }
  }
}

static void DecodeRegOrder(Word Index)
{
  if (!ChkArgCnt(1, 1));
  else if (!DecodeReg(ArgStr[1].Str, BAsmCode + 0)) WrStrErrorPos(ErrNum_InvReg, &ArgStr[1]);
  else
  {
    BAsmCode[0] |= Index; CodeLen = 1;
  }
}

static void DecodeMem(Word Index)
{
  if (ChkArgCnt(1, 1))
  if (DecodeAdr(&ArgStr[1], True, 0, BAsmCode + 0))
  {
    BAsmCode[0] |= Index; CodeLen = 2; ChkPage();
  }
}

static void DecodeJmp(Word Index)
{
  if (ChkArgCnt(1, 1))
  if (DecodeAdr(&ArgStr[1], False, 1, BAsmCode + 0))
  {
    BAsmCode[0] |= Index; CodeLen = 2; ChkPage();
  }
}

static void DecodeLD(Word Index)
{
  if (ChkArgCnt(1, 1))
  if (DecodeAdr(&ArgStr[1], False, 0, BAsmCode + 0))
  {
    BAsmCode[0] |= Index; CodeLen = 2; ChkPage();
  }
}

/*---------------------------------------------------------------------------*/

static void AddFixed(const char *NName, Byte NCode)
{
  AddInstTable(InstTable, NName, NCode, DecodeFixed);
}

static void AddImm(const char *NName, Byte NCode)
{
  AddInstTable(InstTable, NName, NCode, DecodeImm);
}

static void AddReg(const char *NName, Byte NCode)
{
  AddInstTable(InstTable, NName, NCode, DecodeRegOrder);
}

static void AddMem(const char *NName, Byte NCode)
{
  AddInstTable(InstTable, NName, NCode, DecodeMem);
}

static void AddJmp(const char *NName, Byte NCode)
{
  AddInstTable(InstTable, NName, NCode, DecodeJmp);
}

static void InitFields(void)
{
  InstTable = CreateInstTable(201);

  AddFixed("LDE" ,0x40); AddFixed("XAE" ,0x01); AddFixed("ANE" ,0x50);
  AddFixed("ORE" ,0x58); AddFixed("XRE" ,0x60); AddFixed("DAE" ,0x68);
  AddFixed("ADE" ,0x70); AddFixed("CAE" ,0x78); AddFixed("SIO" ,0x19);
  AddFixed("SR"  ,0x1c); AddFixed("SRL" ,0x1d); AddFixed("RR"  ,0x1e);
  AddFixed("RRL" ,0x1f); AddFixed("HALT",0x00); AddFixed("CCL" ,0x02);
  AddFixed("SCL" ,0x03); AddFixed("DINT",0x04); AddFixed("IEN" ,0x05);
  AddFixed("CSA" ,0x06); AddFixed("CAS" ,0x07); AddFixed("NOP" ,0x08);

  AddImm("LDI" , 0xc4); AddImm("ANI" , 0xd4); AddImm("ORI" , 0xdc);
  AddImm("XRI" , 0xe4); AddImm("DAI" , 0xec); AddImm("ADI" , 0xf4);
  AddImm("CAI" , 0xfc); AddImm("DLY" , 0x8f);

  AddReg("XPAL", 0x30); AddReg("XPAH", 0x34); AddReg("XPPC", 0x3c);

  AddMem("LD"  , 0xc0); AddMem("ST"  , 0xc8); AddMem("AND" , 0xd0);
  AddMem("OR"  , 0xd8); AddMem("XOR" , 0xe0); AddMem("DAD" , 0xe8);
  AddMem("ADD" , 0xf0); AddMem("CAD" , 0xf8);

  AddJmp("JMP" , 0x90); AddJmp("JP"  , 0x94); AddJmp("JZ"  , 0x98);
  AddJmp("JNZ" , 0x9c);

  AddInstTable(InstTable, "ILD", 0xa8, DecodeLD);
  AddInstTable(InstTable, "DLD", 0xb8, DecodeLD);
}

static void DeinitFields(void)
{
  DestroyInstTable(InstTable);
}

/*---------------------------------------------------------------------------*/

static void MakeCode_SCMP(void)
{
  CodeLen = 0; DontPrint = False;

  /* zu ignorierendes */

  if (Memo("")) return;

  /* Pseudoanweisungen */

  if (DecodeIntelPseudo(TargetBigEndian)) return;

  if (!LookupInstTable(InstTable, OpPart.Str))
    WrStrErrorPos(ErrNum_UnknownInstruction, &OpPart);
}

static Boolean IsDef_SCMP(void)
{
  return False;
}

static void InitPass_SCMP(void)
{
  SetFlag(&TargetBigEndian, BigEndianName, False);
}

static void SwitchFrom_SCMP(void)
{
  DeinitFields();
  ClearONOFF();
}

static void SwitchTo_SCMP(void)
{
  TurnWords = False;
  SetIntConstMode(eIntConstModeC);

  PCSymbol = "$"; HeaderID = 0x6e; NOPCode = 0x08;
  DivideChars = ","; HasAttrs = False;

  ValidSegs = 1 << SegCode;
  Grans[SegCode] = 1; ListGrans[SegCode] = 1; SegInits[SegCode] = 0;
  SegLimits[SegCode] = 0xffff;

  MakeCode = MakeCode_SCMP; IsDef = IsDef_SCMP;
  SwitchFrom = SwitchFrom_SCMP; InitFields();

  AddONOFF("BIGENDIAN", &TargetBigEndian, BigEndianName, False);

  QualifyQuote = QualifyQuote_SingleQuoteConstant;
  IntConstModeIBMNoTerm = True;
}

void codescmp_init(void)
{
  CPUSCMP = AddCPU("SC/MP", SwitchTo_SCMP);

  AddInitPassProc(InitPass_SCMP);
}
